system/include/emscripten/val.h

```diff
@@ -385,116 +385,123 @@ class val {
   explicit val(T&& value) {
     using namespace internal;
 
-    typedef internal::BindingType<T> BT;
     WireTypePack<T> argv(std::forward<T>(value));
-    handle = _emval_take_value(internal::TypeID<T>::get(), argv);
+    new (this) val(_emval_take_value(internal::TypeID<T>::get(), argv));
   }
 
-  val() : handle(EM_VAL(internal::_EMVAL_UNDEFINED)) {}
+  val() : val(EM_VAL(internal::_EMVAL_UNDEFINED)) {}
 
   explicit val(const char* v)
-      : handle(internal::_emval_new_cstring(v))
+      : val(internal::_emval_new_cstring(v))
   {}
 
-  val(val&& v) : handle(v.handle) {
-    v.handle = 0;
+  val(val&& v) : handle_(v.handle_)
+#if !defined(NDEBUG) && defined(_REENTRANT)
+      , thread(v.thread)
+#endif
+  {
+    v.handle_ = 0;
   }
 
-  val(const val& v) : handle(v.handle) {
-    internal::_emval_incref(handle);
+  val(const val& v) : handle_(v.handle_)
+#if !defined(NDEBUG) && defined(_REENTRANT)
+      , thread(v.thread)
+#endif
+  {
+    internal::_emval_incref(handle_);
   }
 
   ~val() {
-    internal::_emval_decref(handle);
+    if (handle_) {
+      internal::_emval_decref(handle_);
+      handle_ = 0;
+    }
   }
 
   EM_VAL as_handle() const {
-    return handle;
+#if !defined(NDEBUG) && defined(_REENTRANT)
+    assert(pthread_equal(thread, pthread_self()) && "val accessed from wrong thread");
+#endif
+    return handle_;
   }
 
   val& operator=(val&& v) & {
-    auto v_handle = v.handle;
-    v.handle = 0;
-    if (handle) {
-      internal::_emval_decref(handle);
-    }
-    handle = v_handle;
+    val tmp(std::move(v));
+    this->~val();
+    new (this) val(std::move(tmp));
```

Comment:
Are the destructor/copy/assignment changes here a separate cleanup/fix?  Can/should then land separately/first?

Reply:
I can split them out, but they were useful here in particular because handle assignments got more complicated in this PR.

Reply:
Should I split them out or ok to leave in this PR?

Reply:
I don't feel strongly.    I always prefer smaller PR and I have C++ move/copy operators a bit mysterious, but as you point out maybe this change isn't very useful when there is just a single member in the class ?   If you think its an independently useful change that I would split it out.

Reply:
I think it could make PR easier to review if that was a problem but otherwise it's only useful in context of this PR, yeah.

Reply:
Is this particular 3 line pattern a common way to implement rvalue assignment?  It looks very generic .. is it different the `= default` version?  (Showing my C++ ignorance here).

Reply:
This seems like a doses a bunch of extra incrementing a decrementing of the counter.   I assume tmp here to keep the refcount above zero when the descructor runs?   Why bother doing that?

Reply:
> Is this particular 3 line pattern a common way to implement rvalue assignment? It looks very generic .. is it different the = default version? (Showing my C++ ignorance here).

I'm not sure either... I came up with this particular sequence on my own.

> I assume tmp here to keep the refcount above zero when the descructor runs? Why bother doing that?

Because there is a test that verifies we can do `val = std::move(val);` or something like that (assign value by move to itself) which failed before.

Reply:
More context in https://github.com/emscripten-core/emscripten/pull/20072